openshift_tools/reporting/zabbixplot.py

```diff
@@ -48,45 +64,52 @@ def get_event_data(self):
         triggers = self.zapi.trigger.get(group=self.group, search={"description": self.trigger}, selectHosts="extend")
         triggerids = [t["triggerid"] for t in triggers]
-        self.events = self.zapi.event.get(time_from=self.timestamp_start, time_till=self.timestamp_end,
-                                         triggerids=triggerids, limit=self.limit, selectHosts="extend",
-                                         sortfield="clock", sortorder="DESC")
-
-    def __init__(self, server, username, password, group, trigger, days, limit):
-        self.server = server
-        self.username = username
-        self.password = password
+        self.events = self.zapi.event.get(time_from=self.timespan.timestamp_start,
+                                          time_till=self.timespan.timestamp_end, objectids=triggerids,
+                                          limit=self.limit, selectHosts="extend", sortfield="clock", sortorder="DESC")
+        if len(self.events) >= self.limit:
+            print("Warning! Data has been truncated by the result limit. The graph will be inaccurate.",
+                  file=sys.stderr)
+
+    # Reason: All of these parameters are needed
+    # Status: permanently disabled
+    # pylint: disable=too-many-arguments
+    def __init__(self, zapi, group, trigger, days, limit):
+        self.zapi = zapi
         self.group = group
         self.trigger = trigger
         self.days = days
         self.limit = limit
 
-        self.time_start = datetime.datetime.now() - datetime.timedelta(days=days)
-        self.time_end = datetime.datetime.now() - datetime.timedelta(days=0)
-        self.timestamp_start = int(self.time_start.strftime("%s"))
-        self.timestamp_end = int(self.time_end.strftime("%s"))
-        self.mtime_start = mdates.date2num(self.time_start)
-        self.mtime_end = mdates.date2num(self.time_end)
-
-        self.zapi = ZabbixAPI(self.server)
-        self.zapi.login(self.username, self.password)
+        self.timespan = ZabbixTimespan(days)
         self.events = False
 
-    def plot_aggregate(self, periods, num_hosts):
-        """Plots a bar chart from aggregated event occurences"""
-        self.get_event_data()
-
-        period_count = (periods or self.days) + 1  # One more for the ending pseudo-period
-        periods = np.linspace(self.timestamp_start, self.timestamp_end, num=period_count, endpoint=True)
-        period_seconds = (self.timestamp_end - self.timestamp_start) / (period_count - 1)
-
-        host_func = lambda event: event["hosts"][0]["hostid"]
+    @staticmethod
+    def _calculate_event_downtime(event_value, event_clock, current_period_end, down):
+        """Calculate the amount of downtime or uptime caused by an event"""
+        # TODO: This will ignore events that start before the data
+        if event_value != "0":
+            # Add downtime, assuming this may be the last event
+            deltatime = current_period_end - event_clock
+            down = True
```

Comment:
Consider using early exits for these three cases.

So something like this (untested):
```python
    @staticmethod
    def _calculate_event_downtime(event_value, event_clock, current_period_end, down):
        """Calculate the amount of downtime or uptime caused by an event"""
        # TODO: This will ignore events that start before the data
        if event_value != "0":
            # Add downtime, assuming this may be the last event
            return True, current_period_end - event_clock

        if event_value == "0" and down:
            # Remove downtime, compensating for recovered downtime
            return False, current_period_end - event_clock

        # Unchanged status of "up", and we're not recovering from downtime
        # Last event covered this time, so return a delta of zero
        return down, 0
```